cmd/sonobuoy/app/gen.go

```diff
@@ -229,99 +223,3 @@ func getClient(kubeconfig *Kubeconfig) (*kubernetes.Clientset, error) {
 
 	return client, kubeError
 }
-
-// getConfig generates a config which has the the following rules:
-//  - command line options override config values
-//  - plugins specified manually via flags specifically override plugins implied by mode flag
-//  - config values override default values
-// NOTE: Since it mutates plugin values, it should be called before using them.
-func (g *genFlags) resolveConfig() *config.Config {
-	if g == nil {
-		return config.New()
-	}
-
-	conf := config.New()
-	suppliedConfig := g.sonobuoyConfig.Get()
-	if suppliedConfig != nil {
-		mergeConfigs(suppliedConfig, conf)
-		conf = suppliedConfig
-	}
-
-	// Resolve plugins.
-	//  - If using the plugin flags, no actions needed.
-	//  - Otherwise use the supplied config and mode to figure out the plugins to run.
-	//    This only works for e2e/systemd-logs which are internal plugins so then "Set" them
-	//    as if they were provided on the cmdline.
-	// Gate the logic with a nil check because tests may not specify flags and intend the legacy logic.
-	if g.genflags == nil || !g.genflags.Changed("plugin") {
-		// Use legacy logic; conf.SelectedPlugins or mode if not set
-		if conf.PluginSelections == nil {
-			modeConfig := g.mode.Get()
-			if modeConfig != nil {
-				conf.PluginSelections = modeConfig.Selectors
-			}
-		}
-
-		// Set these values as if the user had requested the defaults.
-		if g.genflags != nil {
-			for _, v := range conf.PluginSelections {
-				g.genflags.Lookup("plugin").Value.Set(v.Name)
-			}
-		}
-	}
-
-	// Have to embed the flagset itself so we can inspect if these fields
-	// have been set explicitly or not on the command line. Otherwise
-	// we fail to properly prioritize command line/config/default values.
-	if g.genflags == nil {
-		return conf
-	}
-
-	if g.genflags.Changed(namespaceFlag) {
-		conf.Namespace = g.namespace
-	}
-
-	if g.genflags.Changed(sonobuoyImageFlag) {
-		conf.WorkerImage = g.sonobuoyImage
-	}
-
-	if g.genflags.Changed(imagePullPolicyFlag) {
-		conf.ImagePullPolicy = g.imagePullPolicy.String()
-	}
-
-	if g.genflags.Changed(timeoutFlag) {
-		conf.Aggregation.TimeoutSeconds = g.timeoutSeconds
-	}
-
-	return conf
-}
-
-func mergeConfigs(dst, src *config.Config) {
-	// Workaround for the fact that an explicitly stated empty slice is still
-	// considered a zero value by mergo. This means that the value given
-	// by the user is not respected. Even a custom transformation can't
-	// get around this. See https://github.com/imdario/mergo/issues/118
-	emptyResources := false
-	if len(dst.Resources) == 0 && dst.Resources != nil {
-		emptyResources = true
-	}
-
-	// Workaround to differentiate a false value and a nil value
-	// Only override dst.Limits.PodLogs.SonobuoyNamespace when it's nil
-	// See https://github.com/imdario/mergo/issues/89
-	var sonobuoyNamespace *bool
-	if dst.Limits.PodLogs.SonobuoyNamespace != nil {
```

Comment:
If not using mergo, this isnt an issue I think

Reply:
Same goes for the others. We start with the default config, load the given config if desired, update only the values they specify, and move on. No weird merging logic any more at all.

The price for this is simply an error (with instructions) if you do something like `--timeout 10 --config custom.json` and you just need to provide the config first so it gets layered correctly. Probably no one will be impacted.